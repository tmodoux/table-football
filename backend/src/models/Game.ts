import { Entity, BaseEntity, PrimaryGeneratedColumn, Column } from "typeorm";
import { ObjectType, Field, ID } from "type-graphql";

@Entity()
@ObjectType()
export class Game extends BaseEntity {
  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id: string;

  @Field(() => String)
  @Column()
  player1: string;

  @Field(() => String)
  @Column()
  player2: string;

  @Field(() => Number)
  @Column()
  goals1: number;

  @Field(() => Number)
  @Column()
  goals2: number;
}
