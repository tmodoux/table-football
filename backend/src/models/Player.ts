import { Entity, BaseEntity, PrimaryGeneratedColumn, Column } from "typeorm";
import { ObjectType, Field, ID } from "type-graphql";

@Entity()
@ObjectType()
export class Player extends BaseEntity {
  @Field(() => ID)
  @PrimaryGeneratedColumn()
  id: string;

  @Field(() => String)
  @Column()
  name: string;

  @Field(() => Number)
  @Column()
  played: number;

  @Field(() => Number)
  @Column()
  wins: number;

  @Field(() => Number)
  @Column()
  losses: number;

  @Field(() => Number)
  @Column()
  goalsFor: number;

  @Field(() => Number)
  @Column()
  goalsAgainst: number;
}
