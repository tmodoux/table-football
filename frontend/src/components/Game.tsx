import { ReactNode, useEffect, useState } from "react";
import { PlayerType } from "../App";
import { Button, InputNumber, List, Row, Space } from "antd";
import { ReactComponent as Player1Icon } from "../icons/player1.svg";
import { ReactComponent as Player2Icon } from "../icons/player2.svg";
import CustomIcon from "./CustomIcon";
import {
  CREATE_GAME,
  END_GAME,
  GET_CURRENT_GAME,
  UPDATE_GAME,
} from "../queries";
import { useLazyQuery, useMutation } from "@apollo/client";
import { MessageInstance } from "antd/es/message/interface";
import CustomAlert from "./CustomAlert";

type GamePropsType = {
  players?: PlayerType[];
  updatePlayers?: (players: PlayerType[]) => void;
  messageApi?: MessageInstance;
  isLive?: boolean;
};

type GameType = {
  id: string;
  player1: string;
  player2: string;
  goals1: number;
  goals2: number;
};

const Game = ({
  players = [],
  updatePlayers = () => {},
  messageApi,
  isLive = false,
}: GamePropsType) => {
  const newGame: GameType = {
    id: "",
    player1: "",
    player2: "",
    goals1: 0,
    goals2: 0,
  };
  const [currentGame, setCurrentGame] = useState<GameType>(newGame);
  const isPlayerSelection = !currentGame.player1 || !currentGame.player2;
  const isGameCreation = isLive && !currentGame.id;
  const isGamePlaying = !!currentGame.id;

  const [getCurrentGame] = useLazyQuery(GET_CURRENT_GAME, {
    onCompleted(data) {
      const currentGame = data.getCurrentGame[0];
      if (currentGame) {
        setCurrentGame({
          id: currentGame.id,
          player1: currentGame.player1,
          player2: currentGame.player2,
          goals1: currentGame.goals1,
          goals2: currentGame.goals2,
        });
        messageApi?.success("Live game successfully loaded!");
      }
    },
    onError(error) {
      messageApi?.error("Error while loading current game!");
      console.log(error);
    },
  });

  const [createGame] = useMutation(CREATE_GAME, {
    onCompleted(data) {
      setCurrentGame({ ...currentGame, id: data.createGame.id });
      messageApi?.success("Game successfully created!");
    },
    onError(error) {
      messageApi?.error("Error while creating game!");
      console.log(error);
    },
    variables: {
      data: { player1: currentGame.player1, player2: currentGame.player2 },
    },
  });

  const [updateGame] = useMutation(UPDATE_GAME, {
    onError(error) {
      messageApi?.error("Error while updating game!");
      console.log(error);
    },
  });

  const [endGame] = useMutation(END_GAME, {
    onCompleted(data) {
      updatePlayers(data.endGame);
      setCurrentGame(newGame);
      messageApi?.success("Scores successfully updated!");
    },
    onError(error) {
      messageApi?.error("Error while ending game!");
      console.log(error);
    },
    variables: { data: currentGame },
  });

  const updateScore = (key: string, value: number | null) => {
    const updatedGame = { ...currentGame, [key]: value || 0 };
    setCurrentGame(updatedGame);
    if (isLive) {
      const { id, ...game } = updatedGame;
      updateGame({
        variables: { id, data: game },
      });
    }
  };

  const clickPlayer = (playerId: string): void => {
    if (playerId === currentGame.player1) {
      setCurrentGame({ ...currentGame, player1: "" });
    } else if (playerId === currentGame.player2) {
      setCurrentGame({ ...currentGame, player2: "" });
    } else if (!currentGame.player1) {
      setCurrentGame({ ...currentGame, player1: playerId });
    } else if (!currentGame.player2) {
      setCurrentGame({ ...currentGame, player2: playerId });
    }
  };

  const getPlayerIcon = (playerId: string): ReactNode => {
    if (playerId !== currentGame.player1 && playerId !== currentGame.player2) {
      return <></>;
    } else {
      return (
        <CustomIcon
          component={
            playerId === currentGame.player1 ? Player1Icon : Player2Icon
          }
        />
      );
    }
  };

  const playersData = players.map((player) => (
    <Button
      className="center-inline"
      disabled={isGamePlaying}
      key={player.id}
      icon={getPlayerIcon(player.id)}
      onClick={() => clickPlayer(player.id)}
    >
      {player.name} ({player.id})
    </Button>
  ));

  useEffect(() => {
    if (isLive) getCurrentGame();
  }, [isLive, getCurrentGame]);

  return (
    <>
      <CustomAlert
        numPlayers={players.length}
        isGameCreation={isGameCreation}
        isGamePlaying={isGamePlaying}
      />
      {players.length ? (
        <Space direction="vertical">
          <b>Players</b>
          <Row>{playersData}</Row>
          {isGameCreation || (
            <>
              <b>Scores</b>
              {[
                { key: "goals1", icon: Player1Icon, value: currentGame.goals1 },
                { key: "goals2", icon: Player2Icon, value: currentGame.goals2 },
              ].map((score) => (
                <InputNumber
                  key={score.key}
                  addonBefore={<CustomIcon component={score.icon} />}
                  value={score.value}
                  min={0}
                  onChange={(value) => updateScore(score.key, value)}
                />
              ))}
            </>
          )}
          <Button
            size="large"
            disabled={isPlayerSelection}
            onClick={() => (isGameCreation ? createGame() : endGame())}
          >
            {isGameCreation ? "Start game" : isLive ? "End game" : "Save game"}
          </Button>
        </Space>
      ) : (
        <></>
      )}
    </>
  );
};

export default Game;
