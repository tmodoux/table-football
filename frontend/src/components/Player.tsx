import { useState } from "react";
import { PlayerType } from "../App";
import { Button, Input, Space, Typography } from "antd";
import { CREATE_PLAYER } from "../queries";
import { useMutation } from "@apollo/client";
import { MessageInstance } from "antd/es/message/interface";

type PlayerPropsType = {
  addPlayer?: (player: PlayerType) => void;
  messageApi?: MessageInstance;
};

const Player = ({ addPlayer = () => { }, messageApi }: PlayerPropsType) => {
  const [name, setName] = useState("");

  const [createPlayer] = useMutation(CREATE_PLAYER, {
    onCompleted(data) {
      addPlayer(data.createPlayer);
      messageApi?.success("Players successfully created!");
    },
    onError(error) {
      messageApi?.error("Error while creating player!");
      console.log(error);
    },
  });

  const createNewPlayer = () => {
    createPlayer({ variables: { data: { name: name } } });
    setName("");
  };

  return (
    <>
      <Typography.Title level={4}>Add a new player</Typography.Title>
      <Space>
        <Input
          placeholder="Enter player name"
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
        <Button
          size="large"
          onClick={() => createNewPlayer()}
          disabled={!name}
        >
          Create new player
        </Button>
      </Space>
    </>
  );
};

export default Player;
