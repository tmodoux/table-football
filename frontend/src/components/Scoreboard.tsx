import { PlayerType } from "../App";
import { Table } from "antd";

type ScoreboardPropsType = {
  players: PlayerType[];
};

const Scoreboard = ({ players }: ScoreboardPropsType) => {
  const columns = [
    {
      title: "Player name",
      dataIndex: "playerName",
    },
    {
      title: "Games played",
      dataIndex: "gamesPlayed",
    },
    {
      title: "Wins",
      dataIndex: "wins",
    },
    {
      title: "Losses",
      dataIndex: "losses",
    },
    {
      title: "Ratio",
      dataIndex: "ratio",
    },
    {
      title: "Goals For",
      dataIndex: "goalsFor",
    },
    {
      title: "Goals Against",
      dataIndex: "goalsAgainst",
    },
    {
      title: "Goals Difference",
      dataIndex: "goalsDifference",
    },
  ];

  const data = players.map((player) => {
    return {
      key: player.id,
      playerName: player.name,
      gamesPlayed: player.played,
      wins: player.wins,
      losses: player.losses,
      ratio: player.wins ? player.played / player.wins : "-",
      goalsFor: player.goalsFor,
      goalsAgainst: player.goalsAgainst,
      goalsDifference: player.goalsFor - player.goalsAgainst,
    };
  });

  return (
    <div>
      <h2>Scoreboard</h2>
      <Table columns={columns} dataSource={data} />
    </div>
  );
};

export default Scoreboard;
