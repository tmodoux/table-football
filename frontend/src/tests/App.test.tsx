import { render } from '@testing-library/react';
import Player from '../components/Player';
import { MockedProvider } from "@apollo/client/testing";
import Game from '../components/Game';
import Scoreboard from '../components/Scoreboard';
import { mockMatchMedia } from './testUtils';

describe('Snapshots tests', () => {
  test('renders App', () => {
    const component =
      <MockedProvider>
        <Player />
      </MockedProvider>;
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  test('renders Game', () => {
    const component =
      <MockedProvider>
        <Game />
      </MockedProvider>;
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  test('renders Player', () => {
    const component =
      <MockedProvider>
        <Player addPlayer={() => { }} />
      </MockedProvider>;
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });

  test('renders Scoreboard', () => {
    mockMatchMedia();
    const component =
      <MockedProvider>
        <Scoreboard />
      </MockedProvider>;
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
});