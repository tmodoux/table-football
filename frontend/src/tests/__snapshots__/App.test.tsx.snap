// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots tests renders App 1`] = `
<div>
  <h4
    class="ant-typography css-dev-only-do-not-override-txh9fw"
  >
    Add a new player
  </h4>
  <div
    class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <input
        class="ant-input css-dev-only-do-not-override-txh9fw"
        placeholder="Enter player name"
        type="text"
        value=""
      />
    </div>
    <div
      class="ant-space-item"
    >
      <button
        class="ant-btn css-dev-only-do-not-override-txh9fw ant-btn-default ant-btn-lg"
        disabled=""
        type="button"
      >
        <span>
          Create new player
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Snapshots tests renders Game 1`] = `
<div>
  <div
    class="ant-alert ant-alert-warning ant-alert-with-description ant-alert-no-icon ant-alert-banner css-dev-only-do-not-override-txh9fw"
    data-show="true"
    role="alert"
  >
    <div
      class="ant-alert-content"
    >
      <div
        class="ant-alert-description"
      >
        You must first create some players using the "Add a new player" form below!
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshots tests renders Player 1`] = `
<div>
  <h4
    class="ant-typography css-dev-only-do-not-override-txh9fw"
  >
    Add a new player
  </h4>
  <div
    class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center"
  >
    <div
      class="ant-space-item"
      style="margin-right: 8px;"
    >
      <input
        class="ant-input css-dev-only-do-not-override-txh9fw"
        placeholder="Enter player name"
        type="text"
        value=""
      />
    </div>
    <div
      class="ant-space-item"
    >
      <button
        class="ant-btn css-dev-only-do-not-override-txh9fw ant-btn-default ant-btn-lg"
        disabled=""
        type="button"
      >
        <span>
          Create new player
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Snapshots tests renders Scoreboard 1`] = `
<div>
  <h2
    class="ant-typography css-dev-only-do-not-override-txh9fw"
  >
    Scoreboard
  </h2>
  <div
    class="ant-alert ant-alert-info ant-alert-with-description ant-alert-no-icon ant-alert-banner css-dev-only-do-not-override-txh9fw"
    data-show="true"
    role="alert"
  >
    <div
      class="ant-alert-content"
    >
      <div
        class="ant-alert-description"
      >
        Players are ranked by Ratio first, then Goals difference, then Goals For, then Player name.
      </div>
    </div>
  </div>
  <div
    class="ant-table-wrapper css-dev-only-do-not-override-txh9fw"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-txh9fw"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-empty"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <thead
                  class="ant-table-thead"
                >
                  <tr>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Player name
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Games played
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Wins
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Losses
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Ratio
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Goals For
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Goals Against
                    </th>
                    <th
                      class="ant-table-cell"
                      scope="col"
                    >
                      Goals Difference
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-placeholder"
                  >
                    <td
                      class="ant-table-cell"
                      colspan="8"
                    >
                      <div
                        class="css-dev-only-do-not-override-txh9fw ant-empty ant-empty-normal"
                      >
                        <div
                          class="ant-empty-image"
                        >
                          <svg
                            height="41"
                            viewBox="0 0 64 41"
                            width="64"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <g
                              fill="none"
                              fill-rule="evenodd"
                              transform="translate(0 1)"
                            >
                              <ellipse
                                cx="32"
                                cy="33"
                                fill="#f5f5f5"
                                rx="32"
                                ry="7"
                              />
                              <g
                                fill-rule="nonzero"
                                stroke="#d9d9d9"
                              >
                                <path
                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                />
                                <path
                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  fill="#fafafa"
                                />
                              </g>
                            </g>
                          </svg>
                        </div>
                        <div
                          class="ant-empty-description"
                        >
                          No data
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
